cibuildwheel/__main__.py

```diff
@@ -147,8 +150,19 @@ def main() -> None:
     options = compute_options(platform=platform, command_line_arguments=args)
 
     package_dir = options.globals.package_dir
-    package_files = {"setup.py", "setup.cfg", "pyproject.toml"}
+    from_sdist = package_dir.is_file()
+    if from_sdist:
+        if not str(package_dir).endswith(".tar.gz"):
+            msg = f"cibuildwheel: Package {package_dir} is not an sdist '.tar.gz' tarball"
+            print(msg, file=sys.stderr)
+            sys.exit(2)
 
+        extract_dir, _, _ = str(package_dir).rpartition(".tar.gz")
+        with tarfile.open(package_dir) as tar:
+            tar.extractall()
```

Comment:
Should with pass a `path=` argument here to extract in a temporary directory instead of the current working directory?

Reply:
The deal for a proper sdist is that the root directory inside the tarball is the base of name of the archive (e.g., minus the tar.gz extension). Exacting in place means that an extract dir will be created alright.

Reply:
@nsoranzo I am fine to create a temp directory but this will then need to be moved back after under the project dir which is the contract for the "package dir" argument so far. I am not sure trying to keep things elsewhere would help and this would require extensive surgery

Reply:
Yes, I think a temp dir is better than extracting inplace. That way cibuildwheel isn't leaking its implementation into the filesystem. But you're right that it's a more substantial change. I'll have a think about the best way to go about this.